boba-chain-ops/node/node.go

```diff
@@ -0,0 +1,143 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon/cmd/rpcdaemon/commands"
+	"github.com/ledgerwatch/erigon/common/hexutil"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/rpc"
+	"github.com/ledgerwatch/log/v3"
+)
+
+type RPC interface {
+	SetJWTAuth(jwtSecret *[32]byte) error
+	SetHeader(key, value string)
+	GetLatestBlock() (*Block, error)
+	GetBlockByNumber(blockNumber *big.Int) (*Block, error)
+	GetTransactionByHash(txHash *common.Hash) (types.Transaction, error)
+	ForkchoiceUpdateV1(fc *commands.ForkChoiceState, attributes *commands.PayloadAttributes) (*ForkchoiceUpdatedResult, error)
+	GetPayloadV1(payloadID *PayloadID) (*commands.ExecutionPayload, error)
+	NewPayloadV1(executionPayload *commands.ExecutionPayload) (*PayloadStatusV1, error)
+}
+
+type BackendRPC struct {
+	client  *rpc.Client
+	timeout time.Duration
+	jwtAuth *JWTAuth
+}
+
+func NewRPC(endpoint string, timeout time.Duration, logger log.Logger) (*BackendRPC, error) {
+	rpc, err := rpc.Dial(endpoint, logger)
```

Comment:
Up to you as to whether you leave the current approach in place or not.  But, given that we're setting JWT auth in a header, it means we must be using the HTTP client, not the WS client.  So, probably a slightly cleaner way to implement this, would be to use the `rpc.DialHTTPWithClient`, and to supply an HTTP client which overrides the Transport to refresh the JWT token as necessary.  It would eliminate the repetitive `RefreshJWTAuth` throughout, and ensure that methods couldn't accidentally forget to invoke it.

That being said, this is a migration path, and I don't expect that we'll be adding lots more built on this client implementation, so, if you'd rather not refactor, I'm okay with that too.

Reply:
Could you please shed some light on `Transport`? I don't have to override it. The geth provides `rpc.WithHTTPAuth` for JWT, but erigon doesn't have this.

Reply:
TLDR; Maybe this comment is best ignored.

You may supply an `*http.Client` to the `rpc.DialHTTPWithClient` call.  This `*http.Client` has a `Transport` field which is an `http.RoundTripper`, which takes an `*http.Request` and returns an `*http.Response`.

You _could_ override this implementation with your own implementation of `RoundTripper` that modifies the request to inject a custom authorization header before passing it through to the standard implementation.  While I quickly poked around during the review, that looked reasonable, but, doing a bit of further reading before replying here, I see that the Golang doc is rather opinionated that this is a bad idea:

```
// ... Similarly, RoundTrip should not attempt to
// handle higher-level protocol details such as redirects,
// authentication, or cookies.
```

So, looks like this probably isn't such a great suggestion after all.

Reply:
Yes, that's what I found out. I thought you have some magic ways to override it. :)